### ISSUE TITLE
TravisCI API target failing, mangle_signature for event 'autodoc-process-signature'

### ISSUE THREAD

**Issue by peterjc** [biopython/biopython#3024]
Seen on #3021 and #3022 

```
Handler <function mangle_signature at 0x7f8ae571b620> for event 'autodoc-process-signature' threw an exception
Makefile:20: recipe for target 'html' failed
```

Might be due to a new release of Sphinx or a dependency?

**Comment by peterjc** [comment]
Last good build from master https://travis-ci.org/github/biopython/biopython/jobs/703668122 has

> alabaster==0.7.12,Babel==2.8.0,certifi==2020.6.20,chardet==3.0.4,docutils==0.16,idna==2.10,imagesize==1.2.0,
> isodate==0.6.0,Jinja2==2.11.2,MarkupSafe==1.1.1,mmtf-python==1.1.2,msgpack==1.0.0,
> mysql-connector-python-rf==2.2.2,numpy==1.19.0,numpydoc==1.0.0,packaging==20.4,Pillow==7.2.0,Pygments==2.6.1,
> pyparsing==2.4.7,pytz==2020.1,rdflib==5.0.0,reportlab==3.5.44,requests==2.24.0,scipy==1.5.0,six==1.15.0,
> snowballstemmer==2.0.0,Sphinx==3.1.1,sphinx-rtd-theme==0.5.0,sphinxcontrib-applehelp==1.0.2,
> sphinxcontrib-devhelp==1.0.2,sphinxcontrib-htmlhelp==1.0.3,sphinxcontrib-jsmath==1.0.1,
> sphinxcontrib-qthelp==1.0.3,sphinxcontrib-serializinghtml==1.1.4,urllib3==1.25.9

Recent failing build https://travis-ci.org/github/biopython/biopython/jobs/704617159 has

> alabaster==0.7.12,Babel==2.8.0,certifi==2020.6.20,chardet==3.0.4,docutils==0.16,idna==2.10,imagesize==1.2.0,
> isodate==0.6.0,Jinja2==2.11.2,MarkupSafe==1.1.1,mmtf-python==1.1.2,msgpack==1.0.0,
> mysql-connector-python-rf==2.2.2,numpy==1.19.0,numpydoc==1.1.0,packaging==20.4,Pillow==7.2.0,Pygments==2.6.1,
> pyparsing==2.4.7,pytz==2020.1,rdflib==5.0.0,reportlab==3.5.44,requests==2.24.0,scipy==1.5.0,six==1.15.0,
> snowballstemmer==2.0.0,Sphinx==3.1.1,sphinx-rtd-theme==0.5.0,sphinxcontrib-applehelp==1.0.2,
> sphinxcontrib-devhelp==1.0.2,sphinxcontrib-htmlhelp==1.0.3,sphinxcontrib-jsmath==1.0.1,
> sphinxcontrib-qthelp==1.0.3,sphinxcontrib-serializinghtml==1.1.4,urllib3==1.25.9

Looks like the change from ``numpydoc==1.0.0`` to ``numpydoc==1.1.0`` is a strong candidate.

**Comment by peterjc** [comment]
Release notes https://numpydoc.readthedocs.io/en/latest/release_notes.html specifically say numpydoc 1.1.0 should fix https://github.com/numpy/numpydoc/issues/268

```
Handler <function mangle_docstrings at 0x7f64b5ba57b8> for event 'autodoc-process-docstring' threw an exception
```

Perhaps in fixing ``mangle_docstrings`` something broke in ``mangle_signature``, or this could be due to the other changes to do with signatures and self. Anyway, we have something to try which will confirm this - downgrading numpydoc.

**Comment by peterjc** [comment]
Seems that an object of type ``Bio.Restriction.Restriction.__text_signature__`` is being passed to the ``_clean_text_signature`` function, new in numpydoc v1.1.0:

https://github.com/numpy/numpydoc/blob/v1.1.0/numpydoc/numpydoc.py#L214

This is frustrating as we've mocked ``Bio.Restriction`` to avoid building docs for it in the first place...

**Comment by peterjc** [comment]
Adding a try/except seems to "fix" it, but something else on my local machine is blocking the complete documentation build:

```python
def _clean_text_signature(sig):
    if sig is None:
        return None
    start_pattern = re.compile(r"^[^(]*\(")
    try:
        start, end = start_pattern.search(sig).span()
    except TypeError:
        # Got a strange looking string...                                                                  
        return None
    start_sig = sig[start:end]
    sig = sig[end:-1]
    sig = re.sub(r'^\$(self|module|type)(,\s|$)','' , sig, count=1)
    sig = re.sub(r'(^|(?<=,\s))/,\s\*', '*', sig, count=1)
    return start_sig + sig + ')'
```

**Comment by peterjc** [comment]
Closed in #3026 by pinning numpydoc
